lib/improver/nbhood/circular_kernel.py

```diff
@@ -172,3 +178,156 @@ def run(self, cube, radius):
+        padshape = np.shape(padded)  # Store size to make unflatten easier
+        padded = padded.flatten()
+        # Add 2nd dimension with each point's neighbourhood points along it
+        nbhood_slices = [
+            np.roll(padded, (padshape[1]*j)+i)
+            for i in range(-ranges_xy[1], ranges_xy[1]+1)
+            for j in range(-ranges_xy[0], ranges_xy[0]+1)
+            if kernel[..., i+ranges_xy[1], j+ranges_xy[0]] > 0.]
+
+        # Collapse this dimension into percentiles (a new 2nd dimension)
+        perc_data = np.percentile(nbhood_slices, self.percentiles, axis=0)
+        # Return to 3D
+        perc_data = perc_data.reshape(
+            len(self.percentiles), padshape[0], padshape[1])
+        # Create a cube for these data:
+        pctcube = self.make_percentile_cube(slice_2d)
+        # And put in data, removing the padding
+        pctcube.data = perc_data[:, ranges_xy[0]:-ranges_xy[0],
+                                 ranges_xy[1]:-ranges_xy[1]]
+        return pctcube
+
+    def run(self, cube, radius):
+        """
+        Method to apply a circular kernel to the data within the input cube in
+        order to derive percentiles over the kernel.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube containing array to apply processing to.
+        radius : Float
+            Radius in metres for use in specifying the number of
+            grid cells used to create a circular neighbourhood.
+
+        Returns
+        -------
+        result : Iris.cube.Cube
+            Cube containing the percentile fields.
+            Has percentile as an added dimension.
+
+        """
+        # Check that the cube has an equal area grid.
+        check_if_grid_is_equal_area(cube)
+        # Take data array and identify X and Y axes indices
+        ranges_tuple = convert_distance_into_number_of_grid_cells(
+            cube, radius, MAX_RADIUS_IN_GRID_CELLS)
+        ranges_xy = np.array(ranges_tuple)
+        kernel = circular_kernel(ranges_xy, ranges_tuple, weighted_mode=False)
+
+        # Loop over each 2D slice to reduce memory demand and derive
+        # percentiles on the kernel. Will return an extra dimension.
+        pctcubelist = iris.cube.CubeList()
+        for slice_2d in cube.slices(['projection_y_coordinate',
+                                     'projection_x_coordinate']):
+            pctcubelist.append(
+                self.pad_and_unpad_cube(slice_2d, kernel))
+        result = pctcubelist.merge_cube()
+        exception_coordinates = (
+            find_dimension_coordinate_mismatch(
+                cube, result, two_way_mismatch=False))
+        result = (
+            check_cube_coordinates(
+                cube, result, exception_coordinates=exception_coordinates))
+
+        # Arrange cube, so that the coordinate order is:
+        # realization, percentile, other coordinates.
+        required_order = []
+        if result.coords("realization"):
```

Comment:
I'm also not sure whether there is any unit tests that directly test this reordering. I may be wrong though and need to check more carefully.

Reply:
This could be factored out. At the moment, it's a bit hard-coded to look at "percentiles_over_neighbourhood". The idea of reordering coordinates at the end of some processing is also done in https://github.com/metoppv/improver/blob/master/lib/improver/spotdata/extract_data.py#L225, and a generic version of this to avoid having multiple utilities that do half a job was the point of #237.

Reply:
Ok, I think we just need to update #237 to include this use case and enough information for us to understand what the use case(s) are for it.